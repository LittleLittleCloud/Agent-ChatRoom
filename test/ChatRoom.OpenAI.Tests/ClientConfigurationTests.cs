using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;
using ApprovalTests;
using ApprovalTests.Approvers;
using ApprovalTests.Core;
using ApprovalTests.Namers;
using ApprovalTests.Reporters;
using ChatRoom.OpenAI;
using ChatRoom.SDK;
using FluentAssertions;
using Json.Schema;
using Json.Schema.Generation;
using Moq;
using Xunit;

namespace ChatRoom.OpenAI.Tests;

public class ClientConfigurationTests
{
    private readonly JsonSerializerOptions _jsonSerializerOptions = new JsonSerializerOptions { WriteIndented = true };
    
    [Fact]
    [UseReporter(typeof(DiffReporter))]
    [UseApprovalSubdirectory("ApprovalTests")]
    public void VerifyConfigurationSchema()
    {
        var schema = new JsonSchemaBuilder()
            .FromType<Configuration>()
            .Build();

        var schemaFileName = "chatroom_openai_configuration_schema.json";
        var schemaFilePath = Path.Join("template", "chatroom.openai", schemaFileName);
        var schemaFile = File.ReadAllText(schemaFilePath);

        var json = JsonSerializer.Serialize(schema, new JsonSerializerOptions { WriteIndented = true });
        Approvals.Verify(json);
        schemaFile.Should().BeEquivalentTo(json);
    }

    [Fact]
    [UseReporter(typeof(DiffReporter))]
    [UseApprovalSubdirectory("ApprovalTests")]
    public void ConfigurationLoadTest()
    {
        var config = new OpenAIAgentConfiguration
        {
            LLMConfiguration = new OpenAIClientConfiguration
            {
                LLMType = LLMType.AOAI,
                OpenAIApiKey = "test",
                OpenAIModelId = "test",
                AzureOpenAIEndpoint = "test",
                AzureOpenAIKey = "test",
                AzureOpenAIDeployName = "test",
                ThirdPartyLLMEndpoint = "test",
                ThirdPartyLLMKey = "test"
            },
            SystemMessage = "test",
            Description = "test",
            Name = "test"
        };

        var json = JsonSerializer.Serialize(config, _jsonSerializerOptions);
        Approvals.Verify(json);
        config = JsonSerializer.Deserialize<OpenAIAgentConfiguration>(json)!;

        json.Should().BeEquivalentTo(JsonSerializer.Serialize(config, _jsonSerializerOptions));
    }
}
