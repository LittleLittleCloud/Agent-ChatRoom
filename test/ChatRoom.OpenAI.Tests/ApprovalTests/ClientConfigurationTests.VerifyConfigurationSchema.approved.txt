{
  "type": "object",
  "properties": {
    "room_config": {
      "type": "object",
      "properties": {
        "room": {
          "$ref": "#/$defs/string"
        },
        "port": {
          "$ref": "#/$defs/integer"
        }
      },
      "description": "The configuration for the chat room"
    },
    "agents": {
      "type": "array",
      "items": {
        "type": "object",
        "properties": {
          "llm_config": {
            "$ref": "#/$defs/openAIClientConfiguration"
          },
          "system_message": {
            "$ref": "#/$defs/string9"
          },
          "agent_description": {
            "$ref": "#/$defs/string10"
          },
          "name": {
            "$ref": "#/$defs/string11"
          }
        },
        "description": "openai agents, the name of agents can\u0027t be duplicated"
      },
      "description": "openai agents, the name of agents can\u0027t be duplicated"
    }
  },
  "$defs": {
    "string": {
      "type": "string",
      "description": "The name of the room. Default is \u0027room\u0027"
    },
    "integer": {
      "type": "integer",
      "description": "The port number where the room is hosted. Default is 30000"
    },
    "openAIClientConfiguration": {
      "type": "object",
      "properties": {
        "llm_type": {
          "$ref": "#/$defs/lLMType"
        },
        "openai_api_key": {
          "$ref": "#/$defs/string1"
        },
        "openai_model_id": {
          "$ref": "#/$defs/string2"
        },
        "azure_openai_endpoint": {
          "$ref": "#/$defs/string3"
        },
        "azure_openai_key": {
          "$ref": "#/$defs/string4"
        },
        "azure_openai_deploy_name": {
          "$ref": "#/$defs/string5"
        },
        "third_party_llm_endpoint": {
          "$ref": "#/$defs/string6"
        },
        "third_party_llm_key": {
          "$ref": "#/$defs/string7"
        },
        "third_party_llm_model_id": {
          "$ref": "#/$defs/string8"
        }
      },
      "description": "llm configuration"
    },
    "lLMType": {
      "enum": [
        "AOAI",
        "OpenAI",
        "ThirdParty"
      ],
      "description": "LLM provider, default is Azure OpenAI"
    },
    "string1": {
      "type": "string",
      "description": "OpenAI API key, default is $env:OPENAI_API_KEY"
    },
    "string2": {
      "type": "string",
      "description": "OpenAI model ID, default is gpt-3.5-turbo"
    },
    "string3": {
      "type": "string",
      "description": "Azure OpenAI endpoint, default is $env:AZURE_OPENAI_ENDPOINT"
    },
    "string4": {
      "type": "string",
      "description": "Azure OpenAI key, default is $env:AZURE_OPENAI_API_KEY"
    },
    "string5": {
      "type": "string",
      "description": "Azure OpenAI deploy name, default is $env:AZURE_OPENAI_DEPLOY_NAME"
    },
    "string6": {
      "type": "string",
      "description": "Third-party LLM provider endpoint, default is $env:THIRD_PARTY_LLM_ENDPOINT"
    },
    "string7": {
      "type": "string",
      "description": "Third-party LLM provider key, default is $env:THIRD_PARTY_LLM_KEY"
    },
    "string8": {
      "type": "string",
      "description": "Third-party LLM provider model ID, default is $env:THIRD_PARTY_LLM_MODEL_ID"
    },
    "string9": {
      "type": "string",
      "description": "System message used in gpt agent, default is \u0027You are a helpful AI assistant\u0027"
    },
    "string10": {
      "type": "string",
      "description": "Agent description used in gpt agent, default is \u0027I am a helpful AI assistant\u0027"
    },
    "string11": {
      "type": "string",
      "description": "Name of the agent, default is \u0027gpt\u0027"
    }
  }
}